services/src/main/java/org/keycloak/authorization/admin/PolicyService.java

```diff
@@ -17,115 +17,56 @@
-import org.eclipse.microprofile.openapi.annotations.media.Schema;
-import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
-import org.eclipse.microprofile.openapi.annotations.responses.APIResponses;
-import org.jboss.resteasy.reactive.NoCache;
 import org.keycloak.authorization.AuthorizationProvider;
 import org.keycloak.authorization.model.Policy;
-import org.keycloak.authorization.model.Resource;
 import org.keycloak.authorization.model.ResourceServer;
-import org.keycloak.authorization.model.Scope;
-import org.keycloak.authorization.policy.provider.PolicyProviderAdminService;
 import org.keycloak.authorization.policy.provider.PolicyProviderFactory;
-import org.keycloak.authorization.store.PolicyStore;
-import org.keycloak.authorization.store.ResourceStore;
-import org.keycloak.authorization.store.ScopeStore;
-import org.keycloak.authorization.store.StoreFactory;
-import org.keycloak.events.admin.OperationType;
-import org.keycloak.events.admin.ResourceType;
-import org.keycloak.models.Constants;
-import org.keycloak.models.KeycloakSession;
-import org.keycloak.models.utils.ModelToRepresentation;
 import org.keycloak.representations.idm.authorization.AbstractPolicyRepresentation;
-import org.keycloak.representations.idm.authorization.PolicyProviderRepresentation;
 import org.keycloak.representations.idm.authorization.PolicyRepresentation;
 import org.keycloak.services.ErrorResponseException;
-import org.keycloak.services.resources.KeycloakOpenAPI;
 import org.keycloak.services.resources.admin.AdminEventBuilder;
 import org.keycloak.services.resources.admin.permissions.AdminPermissionEvaluator;
 import org.keycloak.util.JsonSerialization;
 
 /**
  * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
  */
-@Extension(name = KeycloakOpenAPI.Profiles.ADMIN, value = "")
-public class PolicyService {
-
-    protected final ResourceServer resourceServer;
-    protected final AuthorizationProvider authorization;
-    protected final AdminPermissionEvaluator auth;
-    protected final AdminEventBuilder adminEvent;
-
+public class PolicyService extends BasePolicyService {
     public PolicyService(ResourceServer resourceServer, AuthorizationProvider authorization, AdminPermissionEvaluator auth, AdminEventBuilder adminEvent) {
-        this.resourceServer = resourceServer;
-        this.authorization = authorization;
-        this.auth = auth;
-        this.adminEvent = adminEvent.resource(ResourceType.AUTHORIZATION_POLICY);
+        super(resourceServer, authorization, auth, adminEvent);
     }
 
     @Path("{type}")
-    public Object getResource(@PathParam("type") String type) {
+    public Object getResourceServiceOrTypeService(@PathParam("type") String type) {
+        try {
+            return getTypeService(type);
+        } catch (ErrorResponseException ex) {
+            return getResourceService(type);
+        }
+    }
+
+    @Path("/by-type/{policy-type}")
```

Comment:
I had a look at this PR as well, and I am (and OpenAPI) is struggling with the `Object` return type. By splitting the policy-by-type and policy-by-ID to separate endpoints, it would allow a more specific return type than `Object`. Happy to hear more direction from you about this.

Reply:
Yes, that's exactly why I've introduced these two new endpoints.

Reply:
My main concern is that we now have those two endpoints plus the current one returning an object and that does not work well with Open API. Shall we at least deprecate the current endpoint in favor of these two and remove it in 1 or 2 releases from now?

I was wondering if we could somehow handle this using a `org.eclipse.microprofile.openapi.OASFilter` so that we could add both variants (type/id) to the resulting model tree. But I don't see how we can do that.

Reply:
Open API doesn't really document the current one anyway as it's returning `Object` and thus the generator cannot actually document all the endpoints in those services.
If you're okay with that deprecating would be a good idea but that would necessarily involve changes to the admin UI as well, so I didn't do it right away in this PR.